src/base/iter.rs

```diff
@@ -370,10 +416,11 @@ impl<'a, T, R: Dim, C: Dim, S: 'a + RawStorageMut<T, R, C>> Iterator
     type Item = MatrixSliceMut<'a, T, R, U1, S::RStride, S::CStride>;
 
     #[inline]
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.curr < self.ncols() {
-            let res = unsafe { (*self.mat).column_mut(self.curr) };
-            self.curr += 1;
+    fn next(&'_ mut self) -> Option<Self::Item> {
+        if self.range.start < self.ncols() {
+            let pmat: *mut _ = self.mat;
+            let res = unsafe { (*pmat).column_mut(self.range.start) };
```

Comment:
From a brief look, I suspect both the original impl and the new impl might be UB due to provenance issues. Basically here we're creating a temporary mutable reference to the matrix, creating a derived matrix slice and returning it. Because the internal pointers stored in the matrix slice will have provenance derived from the temporary reference, I don't think this is technically legal. It would be great for someone more knowledgeable in provenance / UB to comment on this.

Reply:
Hey, thanks a lot for taking a look so quickly. I had the same fears concerning the sequential iterator but unfortunately I don't know enough yet to fix it for the general case.

I wonder if I can do that for matrices with contiguous storage by using slices to access the underlying data. I might try implementing a special iterator only for those types of matrices. But there's probably something I'm not seeing that would make it hard again. Also I'd be interested in a general solution.

Reply:
I really habe to read up on the whole provenance business. I can surely change the logic back to the original pointer. I changed that around because that way I tricked the compiler into deriving `Send` for me, but I think I should be fine deriving it unsafely...

From the Rustonomicon:
> Something can safely be Send unless it shares mutable state with something else without enforcing exclusive access to it. [link](https://doc.rust-lang.org/nomicon/send-and-sync.html)

I think the iterator does enforce exclusive access (not to the whole matrix but to the memory locations that can be manipulated). But I'm too tired to reason about that right now :) I'll take a look tomorrow. Let me know what you think, when you have time.

Reply:
I really habe to read up on the whole provenance business. I can surely change the logic back to the original pointer. I changed that around because that way I tricked the compiler into deriving `Send` for me, but I think I should be fine deriving it unsafely...

From the Rustonomicon:
> Something can safely be Send unless it shares mutable state with something else without enforcing exclusive access to it. [link](https://doc.rust-lang.org/nomicon/send-and-sync.html)

I think the iterator does enforce exclusive access (not to the whole matrix but to the memory locations that can be manipulated). But I'm too tired to reason about that right now :) I'll take a look tomorrow. Let me know what you think, when you have time.

Reply:
hey again, I've reverted the iterators back to the original pointer logic and taken a stab at feature gating the rayon functionality.